frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -1570,12 +1573,14 @@ func dispatchArg(d *dispatchState, c *instructions.ArgCommand, metaArgs []instru
 
 		ai := argInfo{definition: arg, location: c.Location()}
 
-		if buildArg.Value != nil {
-			if _, ok := nonEnvArgs[buildArg.Key]; !ok {
-				d.state = d.state.AddEnv(buildArg.Key, *buildArg.Value)
-			}
-			ai.value = *buildArg.Value
+		if buildArg.Value == nil {
+			var blankStr string
```

Comment:
This changes the behavior of `ARG` without default value as described in https://github.com/moby/buildkit/pull/4897#discussion_r1594761441 .

In a Dockerfile like
```
FROM alpine
ARG FOO
ARG BAR=
RUN env && stop
```

The `env` output should show only `BAR=` and no `FOO=`.

To implement this you can check value in `d.buildArgs` before logging error or add the `buildArgs` keys to the values passed to `ProcessWords*`, but you can't add these to the `state.Env`.

Reply:
Ah, that was the perfect example to give - I completely misunderstood what was being said!
Ended up checking `d.buildArgs` rather than passing it into to `ProcessWords`; based on the number of tests it caused to fail, I expect the `ProcessWords` route is more involved. 😂

Reply:
In any case, updated!